src/trs/rewrite.rs

```diff
@@ -0,0 +1,143 @@
+//! Polymorphically typed (Hindley-Milner) First-Order Term Rewriting Systems (no abstraction)
+//!
+//! Much thanks to:
+//! - https://github.com/rob-smallshire/hindley-milner-python
+//! - https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system
+//! - (TAPL; Pierce, 2002, ch. 22)
+
+use polytype::Context as TypeContext;
+use rand::seq::sample_iter;
+use rand::Rng;
+use std::f64::NEG_INFINITY;
+use std::fmt;
+use term_rewriting::trace::Trace;
+use term_rewriting::{Rule, RuleContext, TRS as UntypedTRS};
+
+use super::{Lexicon, ModelParams, SampleError, TypeError};
+
+/// Manages the semantics of a term rewriting system.
+#[derive(Debug, PartialEq, Clone)]
+pub struct TRS {
+    // TODO: may also want to track background knowledge here.
+    pub(crate) lex: Lexicon,
+    // INVARIANT: UntypedTRS.rules ends with lex.background
+    pub(crate) utrs: UntypedTRS,
+    pub(crate) ctx: TypeContext,
+}
+impl TRS {
```

Comment:
I'd also like some examples here, at least for `new`, `add_rule`, and `posterior`.

Reply:
Done!